### PR TITLE
Fix kernel_size != 3 on winograd arm/mali.

### PR DESCRIPTION
**Pull request by cbalint13**
This PR fix/enable cases for kernel_size other than 3x3.

* It is a continuation on #3553 work.
* Fix testcases for arm / mali and also #3641 .

@tmoreau89 , @merrymercy , @kevinthesun , @FrozenGene 
Please help with the review.

------
**Main benefit:**
We have flexible ```tile_size``` and any ```kernel_size``` for winograd.

------

**Examples** (yolov3-tiny) of having **tuned ```tile_size```**:

**```original (only direct)```**
[Task 13/20 (1, 384, 26, 26)|(256, 384, 3, 3)] (conv2d) {53.46 GFLOPS /direct}
[Task 14/20 (1, 512, 13, 13)|(1024, 512, 3, 3)] (conv2d) {31.17 GFLOPS /direct}
[Task 15/20 (1, 256, 13, 13)|(512, 256, 3, 3)] (conv2d) {28.77 GFLOPS /direct}

**```original (no tile_size autotune)```**
[Task 13/20 (1, 384, 26, 26, 'float16')] (conv2d) {93.78 GFLOPS /winograd} ```tile_size=2```
[Task 14/20 (1, 512, 13, 13, 'float16')] (conv2d) {93.31 GFLOPS /winograd} ```tile_size=2```
[Task 15/20 (1, 256, 13, 13, 'float16')] (conv2d) {81.34 GFLOPS /winograd} ```tile_size=2```

**```proposed (with tile_size autotune)```**
[Task 13/20 (1, 384, 26, 26, 'float16')] (conv2d) {98.48 GFLOPS /winograd} ```tile_size=4```
[Task 14/20 (1, 512, 13, 13, 'float16')] (conv2d) {106.15 GFLOPS /winograd} ```tile_size=4```
[Task 15/20 (1, 256, 13, 13, 'float16')] (conv2d) {84.54 GFLOPS /winograd} ```tile_size=4```

It is quite tedious to expose more benefits (WiP on my side) for tuned ```tile_size```, these 3 was found running >1 week on mali. That's all what I personally found up to this day, i am sure there are more, will try soon 5x5 & 7x7 kernels (Unet, SuperRes), and post on tophub exceptional ones.

### CODE REVIEW

**Comment by tmoreau89** [comment]
Is there a reason why we are disabling cuda tests for now?

**Comment by cbalint13** [comment]
> Is there a reason why we are disabling cuda tests for now?

* (kernel_size > 3) was running on cuda only.
* now with this runs on all targeted arch.

(kernel_size > 3) is new for our winograd, since its a larger kernel the accuracy may also drop, interesting that wasn't catched before (i think memoization data was changed).

**Comment by cbalint13** [comment]
Furthermore, 

* Also, in future would like to add ```float16``` too for this testcase, and for that rtol has to be even higher ~1e-4, so i think cannot avoid to handle different rtol values in our winograd conv2d tests.

**Comment by cbalint13** [comment]
@merrymercy , @kevinthesun , @FrozenGene

Will update on [tophub](https://github.com/uwsampl/tvm-distro/tree/master/tophub) .
Allow a little time for tophub update (need to fill ```tile_size```), until then this PR is WiP.

File: topi/python/topi/arm_cpu/conv2d.py [hunk 1]
```diff
@@ -340,8 +343,8 @@ def _decl_winograd(cfg, data, kernel, strides, padding, dilation, layout, out_dt
     K = CO
     C = CI
 
-    H = (IH + 2 * HPAD - 3) // HSTR + 1
-    W = (IW + 2 * WPAD - 3) // WSTR + 1
+    H = (IH + 2 * HPAD - KH) // HSTR + 1
```
Comment (FrozenGene) [inline topi/python/topi/arm_cpu/conv2d.py]: Could you help to change `HPAD` / `WPAD` to `pad_top + pad_down` / `pad_left + pad_right` like `spatial_pack` do? We could get `pad_top / down / left / right` from `get_pad_tuple`. Thanks.

File: topi/python/topi/mali/conv2d.py [hunk 2]
```diff
@@ -188,21 +188,16 @@ def _schedule_spatial_pack(cfg, s, output, conv, data_vec, kernel_vec):
...
 def conv2d_mali_winograd(cfg, data, kernel, strides, padding, dilation, layout, out_dtype):
-    tile_size = _pick_tile_size(data, kernel)
-    return _decl_winograd(cfg, data, kernel, strides, padding, dilation, layout, out_dtype,
-                          tile_size)
+    return _decl_winograd(cfg, data, kernel, strides, padding, dilation, layout, out_dtype)
+
+def _decl_winograd(cfg, data, kernel, strides, padding, dilation, layout, out_dtype,
+                   tile_size=None):
+
+    cfg.define_knob('tile_size', enum_tile_sizes(data))
```
Comment (FrozenGene) [inline topi/python/topi/mali/conv2d.py]: Maybe we should upgrade our `tune_relay_arm` / `tune_relay_mali` tutorial, we should make xgb tuner feature_type be `knob`. Otherwise, we maybe hang forever.

Comment (cbalint13) [reply topi/python/topi/mali/conv2d.py]: @FrozenGene ,

 Are you sure about ```xgb``` vs. ```xgb_knob``` ? Cfg space is increasing only few times, its under one milion in size. In my experience ```xgb``` is still performing way better, in fact ```xgb_knob``` did not found anything useful. But I may wrong, please confirm / decline .

Comment (FrozenGene) [reply topi/python/topi/mali/conv2d.py]: `XGBTunner`'s default `feature_type` is `itervar`, see: https://github.com/dmlc/tvm/blob/master/python/tvm/autotvm/tuner/xgboost_tuner.py#L68. When we have `define_knob`, if we don't change `feature_type`from `itervar` into `knob`, we maybe will meet hang forever (i.e. can not tune and just hang). Not related with config search space. Just our feature exaction issue. I met this problem when I `define_knob` in depthwise convolution optimization on ARM CPU. cc @merrymercy

File: topi/python/topi/arm_cpu/conv2d.py (hunk 1 above)
Comment (cbalint13) [reply topi/python/topi/arm_cpu/conv2d.py]: @FrozenGene ,
It is done. Please help double check for mistakes, i did it in quite many places, and altered more than that in cuda's winograd (added ```dilation``` related assertion).

File: topi/python/topi/mali/conv2d.py (hunk 2 above)
Comment (cbalint13) [reply topi/python/topi/mali/conv2d.py]: @FrozenGene ,

 * Experienced those hangouts even crashes when cfg.space was huge like in #2111 (working on it). My feeling was that features get corrupted bad way and xgb gets confused (too large !?). Willing to fix that issue anyway i am working on bring #2111 in TVM finally in (hope to look/fix at ```xgb``` issue too). My test (so far) shows that too much amount of (options like knob) may cause hangs.
  * On other side, we already have knobs there (working fine) we just added another one.

 I triple check: still sure want to update related tutorials to ```xgb_knob``` ? I am confident it works just fine with plain ```xgb``` (and better), but if you still want, I firmly update tutorials (immediate, no more objections).

Comment (FrozenGene) [reply topi/python/topi/mali/conv2d.py]: Ok. Let's us keep `xgb`. Wish you bring #2111 in could have time to look at `xgb` issue too. :-)

Comment (merrymercy) [reply topi/python/topi/mali/conv2d.py]: XGBTuner with `feature_type="itervar"` has the restriction that the AST tree should keep the same structure for all schedule in the search space. So we can get the fix-length feature vector for all schedule in the search space.

When I implemented the winograd schedule, I found different tile size will produce feature vectors with different lengths, so I didn't include it as a tunable knob.  

I expect the tuning will hang as @FrozenGene said, but it seems your change works well.. I will investigate later.

File: topi/python/topi/arm_cpu/conv2d.py (hunk 1 above)
Comment (FrozenGene) [reply topi/python/topi/arm_cpu/conv2d.py]: Could you help to explain why we add `dilation` assertion in CUDA? For CUDA, better way maybe should like we do in ARM CPU / Mali / nn.conv2d_nchw (that is another story), not just `dialte` kernel directly.

File: topi/python/topi/mali/conv2d.py (hunk 2 above)
Comment (cbalint13) [reply topi/python/topi/mali/conv2d.py]: @merrymercy ,
  That was my finding too for the ```xgb```, but it was triggered only by #2111 . Will look into depth at that time, and will open a discussion subject on this topic if will be necessary.
  I don't touch tutorials for this work.

File: topi/python/topi/arm_cpu/conv2d.py (hunk 1 above)
Comment (cbalint13) [reply topi/python/topi/arm_cpu/conv2d.py]: @FrozenGene ,

 Removed the inserted```assert``` . Still concerned if is good that ```HPAD = WPAD = 1``` is removed from ```pre-transformed``` case handler.
 
Can tell if padding related lines are fine now for cuda ?

File: topi/python/topi/nn/winograd_util.py [hunk 3]
```diff
@@ -149,3 +151,16 @@ def winograd_transform_matrices(tile_size, kernel_size, out_dtype):
...
     )
+
+def enum_tile_sizes(data):
+    """Propose tile sizes for winograd convolution"""
+    _, _, H, _ = get_const_tuple(data.shape)
+
+    sizes = [2,]
+    # 2 always present
+    for s in range(3, 9):
+    # overlap less then half
```
Comment (cbalint13) [inline topi/python/topi/nn/winograd_util.py]: @merrymercy , @FrozenGene ,
What is your opinion on this approach to construct ```tile_size``` ?

**Comment by cbalint13** [comment]
> @merrymercy , @kevinthesun , @FrozenGene
> 
> Will update on [tophub](https://github.com/uwsampl/tvm-distro/tree/master/tophub) .
> Allow a little time for tophub update (need to fill `tile_size`), until then this PR is WiP.

It's done, review ready at https://github.com/uwsampl/tvm-distro/pull/8 .

File: topi/python/topi/arm_cpu/conv2d.py (hunk 1 above)
Comment (cbalint13) [reply topi/python/topi/arm_cpu/conv2d.py]: @FrozenGene ,

 Ok, not touching ```dilation``` related anymore, maybe will subject for another PR. I reverted this part from cuda/winograd . Let's remain only at ```tile_size``` introduction for ```cuda``` .
 Please confirm / reslove-conversation if find no problems.

Comment (FrozenGene) [reply topi/python/topi/arm_cpu/conv2d.py]: Yes. Thanks. We could do `dilation` and `padding` in another PR. Make sense.

File: topi/python/topi/nn/winograd_util.py (hunk 3 above)
Comment (FrozenGene) [reply topi/python/topi/nn/winograd_util.py]: I have no strong preference what way should we do. If we think it is a bit trick, we could add comment explaining what happened here, then I think it is enough.

File: python/tvm/autotvm/tophub.py [hunk 4]
```diff
@@ -36,13 +36,13 @@
 
 # the version of each package
 PACKAGE_VERSION = {
-    'arm_cpu': "v0.04",
+    'arm_cpu': "v0.05",
     'llvm':    "v0.03",
 
-    'cuda':    "v0.04",
+    'cuda':    "v0.05",
```
Comment (tmoreau89) [inline python/tvm/autotvm/tophub.py]: CI test with "fail to download tophub package". Have the TOPHUB schedules for cuda been updated?

Comment (tmoreau89) [reply python/tvm/autotvm/tophub.py]: see http://ci.tvm.ai:8080/blue/organizations/jenkins/tvm/detail/PR-3642/20/pipeline/240

**Comment by merrymercy** [comment]
Aha... I remember why I don't include `tile_size` as a tunable knob. This is not the `xgb` vs `xgb_knob` problem. It is because the search spaces for different tile sizes are different. But AutoTVM does not support this kind of space definition.

There are some knobs for tiling in the search space. However, they are dependent on tile size. For example, in the arm cpu's schedule, https://github.com/dmlc/tvm/blob/36702a7678573fa4ff9ea04c3920e60d590496c9/topi/python/topi/arm_cpu/conv2d.py#L390-L392. `P` is the number of winograd tiles. In one case, P is 196 if tile size is 4 while is 784 if tile size is 2. We construct the split knob according the the lengths of P, so different tile sizes will have different spaces.

Currently, AutoTVM does not support dependent relationship between tuning knobs. What it does is picking the first candidate of tile size (which is 2 according your construction) and initializing other knobs. In the above case it will enumerate all split schemes for P = 784. But some split schemes for P = 784 are invalid for P = 196, which will cause the AutoTVM hangs forever during feature extraction.

### Ideal solution
Support hierarchical space definition in AutoTVM

### Workaround
1. Fix the feature extraction. We hang on this line https://github.com/dmlc/tvm/blob/11da1ca31b8121e9a40c70f33ade27670d22e2d0/python/tvm/autotvm/tuner/xgboost_cost_model.py#L311 because we get exception when run this function https://github.com/dmlc/tvm/blob/11da1ca31b8121e9a40c70f33ade27670d22e2d0/python/tvm/autotvm/tuner/xgboost_cost_model.py#L328
2. Use `feature_type` = `knob` in the XGBTuner. I won't merge this solution to master.

You don't meet the problem when tuning for CUDA and mali because I added some restrictions to their schedule tempaltes and luckily they work around this problem.

File: topi/python/topi/nn/winograd_util.py (hunk 3 above)
Comment (merrymercy) [reply topi/python/topi/nn/winograd_util.py]: I don't have experiences tuning all these tile sizes. 
For H = 56, it will return [3, 4, 5, 6, 7, 8]. It will be good if you can try to prune some candidates that are very likely to be the optimal ones according to your tuning results.

**Comment by cbalint13** [comment]
@merrymercy ,

> Aha... I remember why I don't include `tile_size` as a tunable knob. This is not the `xgb` vs `xgb_knob` problem. It is because the search spaces for different tile sizes are different. But AutoTVM does not support this kind of space definition.

1. I was driven by the idea that too simple formula like (tile_size = 4 if H % 4 == 0 else 2) can be rather a simple knob with a small list. Now we afford it with flexible winograd matrix computation.
2.  As result fact, I posted in the header here, knob found better solution overriding mentioned static formula (2 vs 4 tile_size).

> 
> There are some knobs for tiling in the search space. However, they are dependent on tile size. For example, in the arm cpu's schedule,
> 
> https://github.com/dmlc/tvm/blob/36702a7678573fa4ff9ea04c3920e60d590496c9/topi/python/topi/arm_cpu/conv2d.py#L390-L392
> 
> . `P` is the number of winograd tiles. In one case, P is 196 if tile size is 4 while is 784 if tile size is 2. We construct the split knob according to the lengths of P, so different tile sizes will have different spaces.

* I also found this correlation and dependency to P in case of our arm/mali.

* Yes i agree, that is also my finding during tests regarding P. Would be great help me elaborate an exact relationship ```tile_size -> P``` . Otherwise, if we don't narrow down the possible lists,  we just can enumerate bulk ```range(0,256)``` for some of knob's there and let xgb find it out but it wold have to search in 10e+7 cfg.space than.

* Again, as fact, even ignoring this dependency for now we are still better than existing formula.
* In fact, for now, solutions with  this knob will be 2 or 4 (very rare 3 hits) in most cases.

> Currently, AutoTVM does not support dependent relationships between tuning knobs. What it does is picking the first candidate of tile size (which is 2 according to your construction) and initializing other knobs. In the above case, it will enumerate all split schemes for P = 784. But some split schemes for P = 784 are invalid for P = 196, which will cause the AutoTVM hangs forever during feature extraction.

* So, could help me find the relation of P given tile_size ?

> 
> ### Ideal solution
> Support hierarchical space definition in AutoTVM
> 
> ### Workaround
> 1. Fix the feature extraction. We hang on this line https://github.com/dmlc/tvm/blob/11da1ca31b8121e9a40c70f33ade27670d22e2d0/python/tvm/autotvm/tuner/xgboost_cost_model.py#L311
>     because we get exceptions when running this function https://github.com/dmlc/tvm/blob/11da1ca31b8121e9a40c70f33ade27670d22e2d0/python/tvm/autotvm/tuner/xgboost_cost_model.py#L328
> 2. Use `feature_type` = `knob` in the XGBTuner. I won't merge this solution to master.

* OK, i put xgb issue on higher pripority, it is trigerred 100% on such scenarios like #2111 

> 
> You don't meet the problem when tuning for CUDA and mali because I added some restrictions to their schedule templates and luckily they work around this problem. But for arm cpu we will meet this problem.

* Where are those restrictions exactly ?

> 
> Let me know if you want to try to fix it or want more information from me. Otherwise, I can fix it later.

Yes i give it a try to fix it, I'll update you on progress.

-----------

To sum up lets settle for now the order :

0. For now, as mentioned & tested,  our PR is still **better** than existing ```two cased conditional```. 
1. Try find xgb fea issue and repair it.
2. Lets see a smarter P given tile_size (please help me elaborate a generating formula).
3. If get ```tile_size->P``` we can try introduce hierarchical knob mode (i have an simple idea for that).


Arguing by point 0. , the PR would be mergeable, otherwise we transform this PR in a long multi PR.
Agree 1 +  2 + 3 are mandatory, i would help tackle each in upcoming another 3 PR (there will be too much debate if we keep in a single PR like this one).

What you think ?

----------

**Comment by cbalint13** [comment]
Opened a [disscuss topic](https://discuss.tvm.ai/t/winograd-variable-tile-size-knob/3565) and started first with ```xgb``` & ```hierarchical knob``` issue.

File: topi/python/topi/nn/winograd_util.py (hunk 3 above)
Comment (cbalint13) [reply topi/python/topi/nn/winograd_util.py]: @merrymercy ,

*  Inappropriate candidates will be dropped out early by ```xgb```, they don't add overhead or lengthen the search time by much. There can be surprises especially picking e.g. tile_size = 3 sometimes.
* Actual code cover well 2 & 4 cases (tests shows also 3 have good results). But we can be more precise only if narrow down space of ```P``` in accord with ```tile_size``` as you mentioned, or let xgb choose its way.

**Comment by merrymercy** [comment]
#3687 #3689 are necessary to merge this PR, otherwise this PR will make tuning on arm cpu fails.

I can merge this after the above two are merged.

**Comment by cbalint13** [comment]
> #3687 #3689 are necessary to merge this PR, otherwise this PR will make tuning on arm cpu fails.
> 
> I can merge this after the above two are merged.



> #3687 #3689 are necessary to merge this PR. Otherwise this PR will make tuning on arm cpu fails.
> 
> I can merge this after the above two are merged.

@merrymercy ,

May for  ```mali```expand other four vars in this PR , namely```yt```, ```t1```,``` t2``` and ```unroll``` [see here](https://discuss.tvm.ai/t/winograd-variable-tile-size-knob/3565/6?u=cbalint13) ?
Or lets do in other PR where I can show up with retuned/enhanced TOPI entries with comparing in numbers ?

**Comment by merrymercy** [comment]
Currently, the spaces of `P` and other knobs that are dependent on `tile_size` are created with `tile_size=2`
Should we modify these knobs to be the union of all valid value or wait until you implement the conditional knob?

**Comment by cbalint13** [comment]
> Currently, the spaces of `P` and other knobs that are dependent on `tile_size` are created with `tile_size=2`
> Should we modify the spaces of these knobs to be the union of all valid values or wait until you implement the conditional knob?

* I would implement conditional/hierarchical knob,  partially done, need to validate/test. Allow a little more time, will post and reference it here.

**Comment by tmoreau89** [comment]
Note that  #3717 is also proposing to update CUDA TOPHUB version

**Comment by tqchen** [comment]
ping @cbalint13 can you update the PR?

**Comment by tqchen** [comment]
ping @cbalint13

**Comment by cbalint13** [comment]
* I close this for now, in order to complete the following is needed
  
  1. Conditinal-Knob , a knob that hierarchicaly depend on other knob
       - implies that list's combinatoric space will be not fixed (not an easy way, triky to implement)
  2. Once such a knob is implement revise winograd schedule schema for <>3, in this very PR.
